src/connector/src/sink/file_sink/batching_log_sink.rs

```diff
@@ -0,0 +1,168 @@
+                    _ => unreachable!(
+                        "update vnode bitmap can be accepted only right after \
+                    barrier, but current state is {:?}",
+                        state
+                    ),
+                }
+                continue;
+            }
+            // begin_epoch when not previously began
+            state = match state {
+                LogConsumerState::Uninitialized => LogConsumerState::EpochBegun {
+                    curr_epoch: epoch,
+                    max_uncommitted_epoch: None,
+                },
+                LogConsumerState::EpochBegun {
+                    curr_epoch,
+                    max_uncommitted_epoch,
+                } => {
+                    assert!(
+                        epoch >= curr_epoch,
+                        "new epoch {} should not be below the current epoch {}",
+                        epoch,
+                        curr_epoch
+                    );
+                    LogConsumerState::EpochBegun {
+                        curr_epoch: epoch,
+                        max_uncommitted_epoch,
+                    }
+                }
+                LogConsumerState::BarrierReceived { prev_epoch } => {
+                    assert!(
+                        epoch > prev_epoch,
+                        "new epoch {} should be greater than prev epoch {}",
+                        epoch,
+                        prev_epoch
+                    );
+                    LogConsumerState::EpochBegun {
+                        curr_epoch: epoch,
+                        max_uncommitted_epoch: Some(prev_epoch),
+                    }
+                }
+            };
+            match item {
+                LogStoreReadItem::StreamChunk { chunk, chunk_id } => {
+                    sink_writer.write_batch(chunk).await?;
+                    match sink_writer.try_commit().await {
+                        Err(e) => {
+                            return Err(e);
+                        }
+                        // The file has been successfully written and is now visible to downstream consumers.
+                        // Truncate the file to remove the specified `chunk_id` and any preceding content.
+                        Ok(true) => {
+                            // If epoch increased, we first need to truncate the previous epoch.
+                            if let Some(max_uncommitted_epoch) = match state {
+                                LogConsumerState::EpochBegun {
+                                    curr_epoch: _,
+                                    max_uncommitted_epoch,
+                                } => max_uncommitted_epoch,
+                                _ => unreachable!("epoch must have begun before handling barrier"),
+                            } {
+                                assert!(epoch > max_uncommitted_epoch);
+                                log_reader.truncate(TruncateOffset::Barrier {
```

Comment:
I think we should reset `max_uncommitted_epoch` to None after truncation on barrier here? Otherwise, we may introduce duplicate truncation on the same barrier.

Reply:
I think there is no correctness issue with duplicate truncation on the same barrier, but agree with resetting `max_uncommitted_epoch`.

Reply:
I am not sure whether truncating the same barrier for multiple times is correct or not so yes it is better to reset it.